GVFS/GVFS.Common/Maintenance/MaintenanceJobManager.cs

```diff
@@ -0,0 +1,341 @@
+﻿using GVFS.Common.FileSystem;
+using GVFS.Common.Git;
+using GVFS.Common.Prefetch;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+
+namespace GVFS.Common.Maintenance
+{
+    public class MaintenanceJobManager : IDisposable

```

Comment:
The word "Manager" is usually a code smell that a class is doing too much, and in fact this class is doing more than just maintenance since it also generates new midx files post-fetch.

And as I noted lower down, we don't really need a concept of a "job" here. Jobs tend to imply that you create some background process that you have a handle to, so that you can check its status, cancel it, check if it is completed, etc. We don't have, or need, any of that here. It's just that some of our work happens to be asynchronous, so let's drop this job concept and simplify the mental model.

All of that being said, I think this is a useful class, and to me it might make sense to keep it as is and just call it the `LocalGVFSCache`. It has one public method for asking it to generate a new midx, and it also runs a background thread to clean itself up as needed. 

I would also move it out of the Maintenance namespace... unless you split it into two classes, one for the active operations on the cache and another for maintenance. I haven't thought too deeply yet about that approach, but that may be a cleaner separation of concerns.

Reply:
The main reason this class has the post-fetch operations and the background prefetch operations is the "stopping" logic, and how we cancel things during unmount. However, it is worth thinking of a better way to remove the operation logic of what a post-fetch is doing or a background prefetch is doing from these classes.

Each needs to have a `Stop()` method (which is a better name than the `TerminateJobs` that I have below).

Reply:
Also: some jobs need a shared file-based lock, but that could be in a separate constants class.

Reply:
I've got some ideas to help separate these concerns, improve testability, and make manual runs easier. I'm going to add them to the design document before I come back to this PR and re-do the refactor.